### DISCUSSION TITLE
cannot use playwright-python in fastapi, got NotImplementedError

### DISCUSSION THREAD

**Discussion by phoenixor** [fastapi/fastapi#8391] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
// api.py
from playwright.async_api import async_playwright
class FS131VIAPI:
    def __init__(self):
        self.cookie = None
    async def fetch(self, params):
        await self.login_for_cookie()
        logger.info(params)

    async def login_for_cookie(self):
        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=False)
            context = await browser.new_context()
            page = await context.new_page()
            await page.goto("https://cn.bing.com/")
            print(await page.title())
            print(await context.cookies())
            self.cookie = await context.cookies()
            await context.close()

// main.py
...
from api import FS131VIAPI
fs131vi_api = FS131VIAPI()
@app.post("/")
async def index():
    await fs131vi_api.fetch({"SERVICE_IDS": "xxx"})
```

### Description

open the browser and call the endpoint '/'
fastapi received the request and call playwright-python to call "https://cn.bing.com/", return cookies

### Operating System

Windows

### Operating System Details

windows 10

### FastAPI Version

0.85.0

### Python Version

3.10.2

### Additional Context

![S)@V _9AB{UFG%U)_732RNB](https://user-images.githubusercontent.com/6568115/193412091-55362d5b-1798-480e-84b7-7377bbb93341.png)

**Comment by iudeen** [comment, marked as answer]
It works fine. I am not able to find any issues. Below is the code I used. 
```python
import fastapi
from playwright.async_api import async_playwright

class FS131VIAPI:
    def __init__(self):
        self.cookie = None
    async def fetch(self, params):
        await self.login_for_cookie()

    async def login_for_cookie(self):
        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=False)
            context = await browser.new_context()
            page = await context.new_page()
            await page.goto("https://cn.bing.com/")
            print(await page.title())
            print(await context.cookies())
            self.cookie = await context.cookies()
            await context.close()


fs131vi_api = FS131VIAPI()
app = fastapi.FastAPI()


@app.get("/")
async def index():
    await fs131vi_api.fetch({"SERVICE_IDS": "xxx"})

```

**Reply by amitsingh-1811** [reply]
hey,
i am doing the same thing but my code is not working. It is giving "[NotImplementedError](url)".

code snippet:
```
from playwright.async_api import async_playwright
from bs4 import BeautifulSoup

class FS131VIAPI:
    def __init__(self):
        self.cookie = None

    async def fetch_page_text(self, url: str):
        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=False)
            context = await browser.new_context()
            page = await context.new_page()
            await page.goto(url, timeout=60000)
            await page.wait_for_timeout(3000)
            html = await page.content()
            self.cookie = await context.cookies()
            await context.close()

            soup = BeautifulSoup(html, "html.parser")
            paragraphs = soup.find_all(["p", "h1", "h2", "h3", "h4", "h5", "h6"])
            text = " ".join(p.get_text(strip=True) for p in paragraphs)
            return text

    async def enrich_results_with_content(self, results, max_chars=2000):
        enriched = []
        for item in results:
            print("item=> ", item)
            content = await self.fetch_page_text(item["link"])
            enriched.append({
                "name": item.get("title", ""),
                "url": item.get("url", ""),
                "content": content
            })
        return enriched
```

and main method is:
```
import sys
import asyncio
import uvicorn
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel
from app.services.scrapper import FS131VIAPI
from app.services.search import search_web
from app.services.llm_client import setup_rag_system, query_documents

if sys.platform.startswith("win"):
    asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())

fs131vi_api = FS131VIAPI()
app = FastAPI()

class QueryRequest(BaseModel):
    query: str

@app.post("/ask")
async def ask_question(request: QueryRequest):
    query = request.query
    try:
        results = search_web(query)

        if not results:
            raise HTTPException(status_code=404, detail="No search results found.")

        enriched_data = await fs131vi_api.enrich_results_with_content(results)
        print("enriched_data=> ",enriched_data[0].get("content"))
        if not enriched_data or not enriched_data[0].get("content"):
            raise HTTPException(status_code=500, detail="Failed to enrich search results.")

        content = enriched_data[0]["content"]
        query_engine = setup_rag_system(content)
        response_data = query_documents(query_engine, query)

        return {
            "query": query,
            "answer": response_data.response,
            "top_results": [item['link'] for item in results[:3]]  # Optional: return top 3 results
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

def main():
    uvicorn.run("main:app", host="127.0.0.1", port=8000, reload=False)

if __name__ == "__main__":
    main()
```

**Comment by NickVeld** [comment]
In order to be able to find your issue more easily could you please replace "NotEmplementedError" with "[NotImplementedError](https://docs.python.org/3.7/library/exceptions.html?highlight=notimplementederror#NotImplementedError)"

**Comment by phoenixor** [comment]
> It works fine. I am not able to find any issues. Below is the code I used.
> 
> ```python
> import fastapi
> from playwright.async_api import async_playwright
> 
> class FS131VIAPI:
>     def __init__(self):
>         self.cookie = None
>     async def fetch(self, params):
>         await self.login_for_cookie()
> 
>     async def login_for_cookie(self):
>         async with async_playwright() as p:
>             browser = await p.chromium.launch(headless=False)
>             context = await browser.new_context()
>             page = await context.new_page()
>             await page.goto("https://cn.bing.com/")
>             print(await page.title())
>             print(await context.cookies())
>             self.cookie = await context.cookies()
>             await context.close()
> 
> 
> fs131vi_api = FS131VIAPI()
> app = fastapi.FastAPI()
> 
> 
> @app.get("/")
> async def index():
>     await fs131vi_api.fetch({"SERVICE_IDS": "xxx"})
> ```

i am running on windows 10, with python 3.10.2, playwright-python 1.26.1, fastapi 0.85.0, which version did you use?

**Comment by pheman** [comment]
I get the same NotImplementedError Error using fastapi and playwright.
windows 10, python 3.10.5, fastapi=0.87.0, playwright=1.28.0

the playwright function is valid just after removing out of the fastapi api wrapper, i.e. remove "@app.get("/test")"

code attached below:
```
import fastapi,uvicorn
from playwright.async_api import async_playwright

class FS131VIAPI:
    def __init__(self):
        self.cookie = None
    async def fetch(self, params):
        await self.login_for_cookie()

    async def login_for_cookie(self):
        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=False)
            context = await browser.new_context()
            page = await context.new_page()
            await page.goto("https://cn.bing.com/")
            print(await page.title())
            print(await context.cookies())
            self.cookie = await context.cookies()
            await context.close()


fs131vi_api = FS131VIAPI()
app = fastapi.FastAPI()


@app.get("/test")
async def index():
    await fs131vi_api.fetch({"SERVICE_IDS": "xxx"})

if __name__ == "__main__":

    uvicorn.run(
        "test:app",
        host="0.0.0.0",
        reload=True,
        port=8002
    )
```

**Comment by pheman** [comment]
problem can be solved by setting reload=False.
more can be found on 
https://github.com/microsoft/playwright-python/issues/1099

```
if __name__ == "__main__":

    uvicorn.run(
        "test:app",
        host="0.0.0.0",
        reload=False,
        port=8002
    )
```

**Reply by yd-mark** [reply]
close reload, it works for me!!!

**Reply by julioccorderoc** [reply]
Works great for me, thanks!

**Reply by anas-rabhi** [reply]
Works, thank you !!!! 👍

**Reply by esr-zh** [reply]
This is the only solution worked for me. Thank you.

**Reply by VGBijpuria** [reply]
Is there a workaround to still reload? Reload is a big boon during development

**Comment by phoenixor** [comment]
> problem can be solved by setting reload=False. more can be found on [microsoft/playwright-python#1099](https://github.com/microsoft/playwright-python/issues/1099)
> 
> ```
> if __name__ == "__main__":
> 
>     uvicorn.run(
>         "test:app",
>         host="0.0.0.0",
>         reload=False,
>         port=8002
>     )
> ```

i am talking about playwright-python work with reload when developing

**Comment by nick-0101** [comment]
@phoenixor did you solve this?

**Reply by jvhuaiying** [reply]
Has this problem been solved by you?

**Comment by jvhuaiying** [comment]
This problem still exists in **2024**

**Reply by kaonick** [reply]
My solution is add below code before unicorn run:

```
asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())
```

**Reply by jvhuaiying** [reply]
> My solution is add below code before unicorn run:
> 
> ```
> asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())
> ```
thank u very much✌️

**Comment by arsh-pv** [comment]
This same problem occurs when used with --workers, any work around for that ?

**Comment by CNSeniorious000** [comment]
Another way to use auto reload with `playwright` is to start `uvicorn` with `watchfiles` cli itself.

Just replace

```sh
uvicorn main:app --reload --....
```

with

```sh
watchfiles "uvicorn main:app --...."
```

And it works almost the same. That's because `uvicorn[standard]` uses `watchfiles` under the hood.

No more os-dependent hack is needed~
